File: TestFSharpADAndroid/TestFSharpADAndroid/MainActivity.fs
namespace TestFSharpADAndroid

open System

open Android.App
open Android.Content
open Android.OS
open Android.Runtime
open Android.Views
open Android.Widget
open System.Linq
open Microsoft.IdentityModel.Clients.ActiveDirectory

type Resources = TestFSharpADAndroid.Resource

[<Activity (Label = "TestFSharpADAndroid", MainLauncher = true)>]
type MainActivity () =
    inherit Activity ()

    let mutable count:int = 1

    override this.OnCreate (bundle) =

        base.OnCreate (bundle)

        // Set our view from the "main" layout resource
        this.SetContentView (Resources.Layout.Main)

        let statusTextView = this.FindViewById<TextView>(Resources.Id.statustextView)

        // Get our button from the layout resource, and attach an event to it
        let button = this.FindViewById<Button>(Resources.Id.connectionButton)

        let AuthenticateAsync(authority, resource, clientId, returnUri) =
            let firstAuthContext = new AuthenticationContext(authority)
            let uri = new Uri(returnUri)

            let authContext =
                match firstAuthContext.TokenCache.ReadItems().Any() with
                | true  -> new AuthenticationContext(firstAuthContext.TokenCache.ReadItems().First().Authority)
                | false -> firstAuthContext

            let activity = this :> Android.App.Activity
            authContext.AcquireTokenAsync(resource, clientId, uri, new PlatformParameters(activity))

        let GetTokenAsync() =
            async {
                try
                    let! result = AuthenticateAsync(GlobalConfiguration.Authority, GlobalConfiguration.AppId, GlobalConfiguration.ClientId, GlobalConfiguration.RedirectUri) |> Async.AwaitTask
                    let ret = match result with
                                | null -> None
                                | _    -> Some(result)
                    return ret
                with
                   | exn -> return None
            }

        let LoginAsync() =
            async {
                let! token = GetTokenAsync()
                let tokenStatus = match token with
                                  | Some t -> true
                                  | None   -> false
                return tokenStatus
            }

        button.Click.Add (fun args -> 
            async {
                let! isAuth = LoginAsync()
                match isAuth with
                | true -> ()
                | false -> ()
            } |> Async.Start
        )

File: TestADAndroid/TestADAndroid/MainActivity.cs
using Android.App;
using Android.Widget;
using Android.OS;
using System.Threading.Tasks;
using Microsoft.IdentityModel.Clients.ActiveDirectory;
using System;
using System.Linq;

namespace TestADAndroid
{
    [Activity(Label = "TestADAndroid", MainLauncher = true)]
    public class MainActivity : Activity
    {
        TextView _statusTextView;

        protected override void OnCreate(Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);

            // Set our view from the "main" layout resource
            SetContentView(Resource.Layout.Main);

            FindViewById<Button>(Resource.Id.connectionButton).Click += connectionButton_Click;
            _statusTextView = FindViewById<TextView>(Resource.Id.statusTextView);
        }

        private async void connectionButton_Click(object sender, System.EventArgs e)
        {
            await LoginAsync();
        }

        public async Task<bool> LoginAsync()
        {
            return (await GetTokenAsync()) != null;
        }

        private async Task<AuthenticationResult> GetTokenAsync()
        {
            try
            {
                var result = await AuthenticateAsync(GlobalConfiguration.Authority, GlobalConfiguration.PrometheeAppId, GlobalConfiguration.ClientId, GlobalConfiguration.RedirectUri);
                if (result != null)
                {
                    _statusTextView.Text = result.UserInfo.ToString();
                    return result;
                }
            }
            catch (Exception) { }

            return null;
        }

        public async Task<AuthenticationResult> AuthenticateAsync(string authority, string resource, string clientId, string returnUri)
        {
            var authContext = new AuthenticationContext(authority);
            if (authContext.TokenCache.ReadItems().Any())
                authContext = new AuthenticationContext(authContext.TokenCache.ReadItems().First().Authority);

            var uri = new Uri(returnUri);
            var activity = (Activity)this;
            var authResult = await authContext.AcquireTokenAsync(resource, clientId, uri, new PlatformParameters(activity));
            return authResult;
        }
    }
}

